authority/provisioner/webhook.go

```diff
@@ -0,0 +1,203 @@
+	"github.com/smallstep/certificates/templates"
+	"go.step.sm/crypto/sshutil"
+	"go.step.sm/crypto/x509util"
+	"golang.org/x/crypto/ssh"
+)
+
+type Webhook struct {
+	ID                   string `json:"id"`
+	Name                 string `json:"name"`
+	URL                  string `json:"url"`
+	Kind                 string `json:"kind"`
+	DisableTLSClientAuth bool   `json:"disableTLSClientAuth,omitempty"`
+	SigningSecret        string `json:"-"`
+	BearerToken          string `json:"-"`
+	BasicAuth            struct {
+		Username string
+		Password string
+	} `json:"-"`
+}
+
+type sshRequest struct {
+	Key        []byte
+	Type       string
+	KeyID      string
+	Principals []string
+}
+
+// ssh.Certificate but the Key and SignatureKey are marshaled
+type sshCert struct {
+	Nonce           []byte          `json:"nonce"`
+	Key             []byte          `json:"key"`
+	Serial          uint64          `json:"serial"`
+	CertType        uint32          `json:"certType"`
+	KeyID           string          `json:"keyID"`
+	ValidPrincipals []string        `json:"validPrincipals"`
+	ValidAfter      uint64          `json:"validAfter"`
+	ValidBefore     uint64          `json:"validBefore"`
+	Permissions     ssh.Permissions `json:"permissions"`
+	Reserved        []byte          `json:"reserved"`
+	SignatureKey    []byte          `json:"signatureKey"`
+	Signature       *ssh.Signature  `json:"Signature"`
+}
+
+type webhookRequestBody struct {
+	Timestamp time.Time `json:"timestamp"`
+	//nolint:revive // acronyms can use all caps
+	X509_CSR []byte `json:"csr,omitempty"`
+	//nolint:revive // acronyms can use all caps
+	SSH_CR         *sshRequest           `json:"ssh_cr,omitempty"`
+	Certificate    *x509util.Certificate `json:"certificate,omitempty"`
+	SSHCertificate *sshCert              `json:"ssh_certificate,omitempty"`
+}
+
+type WebhookResponseBody struct {
+	Data  map[string]interface{} `json:"data"`
+	Allow bool                   `json:"allow"`
+}
+
+func (w *Webhook) Do(ctx context.Context, client *http.Client, certReq interface{}, data map[string]interface{}) (*WebhookResponseBody, error) {
+	tmpl, err := template.New("url").Funcs(templates.StepFuncMap()).Parse(w.URL)
+	if err != nil {
+		return nil, err
+	}
+	buf := &bytes.Buffer{}
+	if err := tmpl.Execute(buf, data); err != nil {
+		return nil, err
+	}
+	url := buf.String()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	retries := 1
+retry:
+	reqBody := &webhookRequestBody{
+		Timestamp: time.Now(),
+	}
+	switch r := certReq.(type) {
+	case *x509.CertificateRequest:
+		if r != nil {
+			reqBody.X509_CSR = r.Raw
+		}
+	case sshutil.CertificateRequest:
+		reqBody.SSH_CR = &sshRequest{
+			Type:       r.Type,
+			KeyID:      r.KeyID,
+			Principals: r.Principals,
+		}
+		if r.Key != nil {
+			reqBody.SSH_CR.Key = r.Key.Marshal()
+		}
+	case *x509.Certificate:
+		reqBody.Certificate = x509util.FromX509Certificate(r)
+	case *ssh.Certificate:
+		reqBody.SSHCertificate = &sshCert{
```

Comment:
It would make sense to use `sshutil.Certificate` here. If we cannot, it's ok. But at least make sure that the types and json tags match.

Reply:
6c6c618ca866b4b1b266497380e3f0dde8f2c7cc